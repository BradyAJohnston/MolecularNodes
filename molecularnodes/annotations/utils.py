import inspect
import types
import typing
import bpy
from PIL import Image


def get_all_class_annotations(cls) -> dict:
    """Get all the annotations from the class including base classes"""
    all_annotations = {}
    for base_cls in cls.mro():
        if hasattr(base_cls, "__annotations__"):
            current_annotations = inspect.get_annotations(base_cls, eval_str=True)
            all_annotations.update(current_annotations)
    if "interface" in all_annotations:
        del all_annotations["interface"]
    return all_annotations


def get_blender_supported_type(atype):
    supported_types = (
        str,
        bool,
        int,
        float,
        tuple[float, float],
        tuple[float, float, float],
        tuple[float, float, float, float],
    )
    if atype in supported_types:
        return atype
    if typing.get_origin(atype) is typing.Union or type(atype) is types.UnionType:
        for utype in atype.__args__:
            if utype in supported_types:
                return utype
    return None


def get_view_matrix(obj):
    if obj._render_mode:
        return obj._scene.camera.matrix_world.inverted()
    else:
        return obj._rv3d.view_matrix


def is_perspective_projection(obj):
    if obj._render_mode:
        return obj._scene.camera.data.type != "ORTHO"
    else:
        return obj._rv3d.is_perspective


def render_annotations(
    scene: bpy.types.Scene, render_scale: float, image: Image, image_scale: float
) -> None:
    """Render annotations of all entities to an image"""
    session = scene.MNSession
    session.prune()  # remove any invalid session entities
    for entity in session.entities.values():
        if not hasattr(entity, "annotations"):
            continue
        manager = entity.annotations
        manager._enable_render_mode(scene, render_scale, image, image_scale)
        manager._draw_annotations()
        manager._disable_render_mode()
