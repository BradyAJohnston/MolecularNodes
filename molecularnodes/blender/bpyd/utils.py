import numpy as np


def centre(position: np.ndarray, weight: np.ndarray | None = None):
    "Calculate the weighted centroid of the vectors"
    if weight is None:
        return np.mean(position, axis=0)
    return np.sum(position * weight.reshape((-1, 1)), axis=0) / np.sum(weight)
