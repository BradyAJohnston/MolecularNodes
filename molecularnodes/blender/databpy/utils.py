import bpy


def evaluate_object(obj: bpy.types.Object):
    "Return an object which has the modifiers evaluated."
    obj.update_tag()
    return obj.evaluated_get(bpy.context.evaluated_depsgraph_get())
