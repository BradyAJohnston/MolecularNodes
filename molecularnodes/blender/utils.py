import bpy
from pathlib import Path


def path_resolve(path: str | Path) -> Path:
    if isinstance(path, str):
        return Path(bpy.path.abspath(path))
    elif isinstance(path, Path):
        return Path(bpy.path.abspath(str(path)))
    else:
        raise ValueError(f"Unable to resolve path: {path}")
