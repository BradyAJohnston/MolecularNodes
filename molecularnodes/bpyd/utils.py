import numpy as np


def centre(position: np.ndarray, weight: np.ndarray | None = None):
    "Calculate the weighted centroid of the vectors"
    if weight is None:
        return np.mean(position, axis=0)
    return np.sum(position * weight.reshape((-1, 1)), axis=0) / np.sum(weight)


def lerp(a: np.ndarray, b: np.ndarray, t: float = 0.5) -> np.ndarray:
    """
    Linearly interpolate between two values.

    Parameters
    ----------
    a : array_like
        The starting value.
    b : array_like
        The ending value.
    t : float, optional
        The interpolation parameter. Default is 0.5.

    Returns
    -------
    array_like
        The interpolated value(s).

    Notes
    -----
    This function performs linear interpolation between `a` and `b` using the
    interpolation parameter `t` such that the result lies between `a` and `b`.

    Examples
    --------
    >>> lerp(1, 2, 0.5)
    1.5

    >>> lerp(3, 7, 0.2)
    3.8

    >>> lerp([1, 2, 3], [4, 5, 6], 0.5)
    array([2.5, 3.5, 4.5])

    """
    return np.add(a, np.multiply(np.subtract(b, a), t))
