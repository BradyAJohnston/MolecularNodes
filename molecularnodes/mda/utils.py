"""
Utils
"""

import bpy
from bpy.app.handlers import persistent  # type: ignore


def is_mda_universe(object: bpy.types.Object) -> bool:
    """Check if blender object is MDA universe"""
    return object and object.mn.mda and object.mn.mda.is_mda_universe


def _msgbus_active_object_callback(*args):
    """Callback when active object changes"""
    sprop = bpy.context.scene.mn.mda
    active_object = bpy.context.active_object
    index = -1
    if is_mda_universe(active_object):
        # find index of universe in scene universes collection
        index = sprop.universes.find(active_object.mn.mda.universe_key)
    sprop.active_index = index  # updates selection in Universes panel


def _subscribe_to_active_object():
    """Actual subscribe to active object changes"""
    subscribe_to = (bpy.types.LayerObjects, "active")
    bpy.msgbus.subscribe_rna(
        key=subscribe_to,
        owner=bpy,
        args=(),
        notify=_msgbus_active_object_callback,
    )


@persistent
def active_object_load_handler(dummy):
    """Load handler to re-subscribe to active object changes"""
    _subscribe_to_active_object()


def subscribe_to_active_object():
    """Subscribe to active object changes"""
    _subscribe_to_active_object()  # subscribe
    # create a load handler to re-subscribe on file loads
    bpy.app.handlers.load_post.append(active_object_load_handler)


def set_object_visibility(object: bpy.types.Object, visible: bool) -> None:
    """Set visibility of Blender object"""
    if object.name not in bpy.context.view_layer.objects:
        return
    try:
        object.hide_set(not visible)
        # obj.hide_viewport = hide
        # icon incorrect, but causes blender bug for volumes with above
        object.hide_render = not visible
    except:
        # Keyframing object visibility can at times lead to:
        # RuntimeError: Object can't be hidden because it is not in View Layer 'ViewLayer'!
        pass
