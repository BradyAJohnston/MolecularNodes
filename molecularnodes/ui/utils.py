import bpy

def check_online_access_for_ui(layout: bpy.types.UILayout) -> bpy.types.UILayout:
    """
    Disable UI without Online Access

    Checks for the online access permissions, and adds a warning and disables following
    UI elements if it fails the check. Returns the UILayout that will have .enabled flag
    set to False, disabling all subsequent uses of the layout.

    Args:
        layout (bpy.types.UILayout): The UILayout element to add the warning and potentially
        disable.

    Returns:
        bpy.types.UILayout: The altered UILayout element, for use in downstream UI
        components.
    """
    if not bpy.app.online_access:
        layout.label(
            text="Online access disabled. Change in Blender's system preferences.",
            icon="ERROR",
        )
        op = layout.operator("wm.url_open", text="Online Access Docs", icon="URL")
        op.url = "https://docs.blender.org/manual/en/dev/editors/preferences/system.html#bpy-types-preferencessystem-use-online-access"
        layout = layout.column()
        layout.alert = True
        layout.enabled = False

    return layout