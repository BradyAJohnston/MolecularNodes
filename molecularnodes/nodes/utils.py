import bpy
import databpy


def get_links(socket: bpy.types.NodeSocket) -> bpy.types.NodeLinks:
    """
    Get the links between two sockets
    """
    links = socket.links
    if links is None:
        raise ValueError("No links found for socket")
    return links


def int_menu_switch_from_dict(
    dictionary: dict, name: str = "NewMenuSwitch"
) -> bpy.types.NodeTree:
    """
    Create a menu from a dictionary of options
    """
    tree = databpy.nodes.new_tree(name=name, geometry=False)

    node_m = tree.nodes.new("GeometryNodeMenuSwitch")
    node_m.data_type = "INT"

    starting_enum_length = len(node_m.enum_items)

    for i, (key, value) in enumerate(dictionary.items()):
        if i < starting_enum_length:
            node_m.enum_items[i].name = key
        else:
            node_m.enum_items.new(name=key)

        node_m.inputs[i + 1].default_value = value

    tree.interface.new_socket("Selector", socket_type="NodeSocketMenu")
    tree.interface.new_socket("Value", socket_type="NodeSocketInt", in_out="OUTPUT")
    tree.links.new(databpy.nodes.get_input(tree).outputs["Selector"], node_m.inputs[0])
    tree.links.new(
        node_m.outputs["Output"], databpy.nodes.get_output(tree).inputs["Value"]
    )
    return tree
