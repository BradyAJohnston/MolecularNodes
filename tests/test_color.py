import molecularnodes as mn
import numpy as np

from .utils import NumpySnapshotExtension


def test_random_rgb(snapshot_custom):
    n = 100
    colors = np.array(list(map(mn.color.random_rgb, range(n))))
    assert snapshot_custom == colors


def test_colos_from_atomic_numbers(snapshot_custom: NumpySnapshotExtension):
    length = len(mn.color.iupac_colors_rgb)
    colors = mn.color.colors_from_elements(np.array(list(range(length))))
    assert snapshot_custom == colors
