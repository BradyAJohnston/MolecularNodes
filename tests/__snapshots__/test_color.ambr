# serializer version: 1
# name: test_colos_from_atomic_numbers
  [[242   0  26   1]
   [255 255 255   1]
   [217 255 255   1]
   [204 128 255   1]
   [194 255   0   1]
   [255 181 181   1]
   [144 144 144   1]
   [ 48  80 248   1]
   [255  13  13   1]
   [144 224  80   1]
   [179 227 245   1]
   [171  92 242   1]
   [138 255   0   1]
   [191 166 166   1]
   [240 200 160   1]
   [255 128   0   1]
   [255 255  48   1]
   [ 31 240  31   1]
   [143  64 212   1]
   [128 209 227   1]
   [ 61 255   0   1]
   [230 230 230   1]
   [191 194 199   1]
   [166 166 171   1]
   [138 153 199   1]
   [156 122 199   1]
   [224 102  51   1]
   [199 138 138   1]
   [255 217 143   1]
   [200 128  51   1]
   [125 128 176   1]
   [194 143 143   1]
   [102 143 143   1]
   [189 128 227   1]
   [255 161   0   1]
   [166  41  41   1]
   [ 92 184 209   1]
   [112  46 176   1]
   [  0 255   0   1]
   [148 255 255   1]
   [148 224 224   1]
   [115 194 201   1]
   [ 84 181 181   1]
   [ 59 158 158   1]
   [ 36 125 125   1]
   [ 10 125 140   1]
   [  0 105 133   1]
   [192 192 192   1]
   [255 217 143   1]
   [166 117 115   1]
   [102 128 128   1]
   [158  99 181   1]
   [212 122   0   1]
   [148   0 148   1]
   [ 66 158 176   1]
   [ 87  23 143   1]
   [  0 201   0   1]
   [112 212 255   1]
   [255 255 199   1]
   [217 255 199   1]
   [199 255 199   1]
   [163 255 199   1]
   [143 255 199   1]
   [ 97 255 199   1]
   [ 69 255 199   1]
   [ 48 255 199   1]
   [ 31 255 199   1]
   [  0 255 156   1]
   [  0 230 117   1]
   [  0 212  82   1]
   [  0 191  56   1]
   [  0 171  36   1]
   [ 77 194 255   1]
   [ 77 166 255   1]
   [ 33 148 214   1]
   [ 38 125 171   1]
   [ 38 102 150   1]
   [ 23  84 135   1]
   [208 208 224   1]
   [255 209  35   1]
   [184 184 208   1]
   [166  84  77   1]
   [ 87  89  97   1]
   [158  79 181   1]
   [255 161   0   1]
   [255 161   0   1]
   [255 161   0   1]
   [255 161   0   1]
   [255 161   0   1]
   [255 161   0   1]
   [255 161   0   1]
   [255 161   0   1]
   [255 161   0   1]
   [255 161   0   1]
   [255 161   0   1]
   [255 161   0   1]
   [255 161   0   1]
   [255 161   0   1]
   [204   0  89   1]
   [209   0  79   1]
   [217   0  69   1]
   [224   0  56   1]
   [230   0  46   1]
   [235   0  38   1]
   [240   0  33   1]
   [241   0  30   1]
   [242   0  26   1]
   [242   0  26   1]
   [242   0  26   1]
   [242   0  26   1]
   [242   0  26   1]
   [242   0  26   1]]
# ---
# name: test_random_rgb
  [[0.8 0.4 0.4 1.0]
   [0.8 0.7 0.4 1.0]
   [0.8 0.4 0.5 1.0]
   [0.6 0.8 0.4 1.0]
   [0.6 0.8 0.4 1.0]
   [0.4 0.5 0.8 1.0]
   [0.7 0.4 0.8 1.0]
   [0.4 0.8 0.4 1.0]
   [0.7 0.8 0.4 1.0]
   [0.4 0.8 0.7 1.0]
   [0.4 0.6 0.8 1.0]
   [0.4 0.8 0.7 1.0]
   [0.4 0.8 0.8 1.0]
   [0.6 0.8 0.4 1.0]
   [0.8 0.7 0.4 1.0]
   [0.8 0.4 0.5 1.0]
   [0.4 0.8 0.4 1.0]
   [0.4 0.8 0.8 1.0]
   [0.8 0.8 0.4 1.0]
   [0.4 0.4 0.8 1.0]
   [0.8 0.4 0.6 1.0]
   [0.8 0.8 0.4 1.0]
   [0.8 0.4 0.5 1.0]
   [0.8 0.4 0.6 1.0]
   [0.5 0.4 0.8 1.0]
   [0.4 0.8 0.5 1.0]
   [0.6 0.4 0.8 1.0]
   [0.4 0.4 0.8 1.0]
   [0.8 0.7 0.4 1.0]
   [0.4 0.7 0.8 1.0]
   [0.4 0.7 0.8 1.0]
   [0.8 0.4 0.4 1.0]
   [0.8 0.6 0.4 1.0]
   [0.4 0.6 0.8 1.0]
   [0.4 0.8 0.8 1.0]
   [0.4 0.7 0.8 1.0]
   [0.4 0.8 0.4 1.0]
   [0.4 0.4 0.8 1.0]
   [0.4 0.4 0.8 1.0]
   [0.7 0.8 0.4 1.0]
   [0.4 0.8 0.7 1.0]
   [0.4 0.8 0.5 1.0]
   [0.4 0.4 0.8 1.0]
   [0.8 0.5 0.4 1.0]
   [0.4 0.8 0.6 1.0]
   [0.5 0.8 0.4 1.0]
   [0.8 0.4 0.7 1.0]
   [0.4 0.8 0.4 1.0]
   [0.4 0.7 0.8 1.0]
   [0.8 0.6 0.4 1.0]
   [0.4 0.8 0.8 1.0]
   [0.6 0.8 0.4 1.0]
   [0.8 0.4 0.4 1.0]
   [0.4 0.5 0.8 1.0]
   [0.8 0.4 0.6 1.0]
   [0.8 0.6 0.4 1.0]
   [0.8 0.4 0.5 1.0]
   [0.8 0.5 0.4 1.0]
   [0.4 0.6 0.8 1.0]
   [0.7 0.8 0.4 1.0]
   [0.4 0.8 0.4 1.0]
   [0.4 0.8 0.8 1.0]
   [0.8 0.4 0.6 1.0]
   [0.4 0.8 0.7 1.0]
   [0.4 0.8 0.8 1.0]
   [0.4 0.8 0.6 1.0]
   [0.8 0.6 0.4 1.0]
   [0.8 0.6 0.4 1.0]
   [0.6 0.4 0.8 1.0]
   [0.4 0.4 0.8 1.0]
   [0.8 0.4 0.6 1.0]
   [0.4 0.8 0.4 1.0]
   [0.8 0.6 0.4 1.0]
   [0.5 0.8 0.4 1.0]
   [0.8 0.4 0.8 1.0]
   [0.4 0.8 0.7 1.0]
   [0.4 0.8 0.5 1.0]
   [0.7 0.4 0.8 1.0]
   [0.8 0.4 0.8 1.0]
   [0.8 0.8 0.4 1.0]
   [0.5 0.8 0.4 1.0]
   [0.4 0.8 0.8 1.0]
   [0.8 0.8 0.4 1.0]
   [0.4 0.8 0.8 1.0]
   [0.5 0.4 0.8 1.0]
   [0.7 0.8 0.4 1.0]
   [0.7 0.4 0.8 1.0]
   [0.8 0.8 0.4 1.0]
   [0.4 0.8 0.5 1.0]
   [0.8 0.6 0.4 1.0]
   [0.7 0.8 0.4 1.0]
   [0.8 0.6 0.4 1.0]
   [0.4 0.8 0.6 1.0]
   [0.8 0.4 0.6 1.0]
   [0.4 0.7 0.8 1.0]
   [0.6 0.4 0.8 1.0]
   [0.4 0.8 0.5 1.0]
   [0.8 0.8 0.4 1.0]
   [0.4 0.8 0.4 1.0]
   [0.4 0.8 0.6 1.0]]
# ---
